extension/js/common/api/email-provider/gmail/google.ts

```diff
@@ -17,69 +18,73 @@ export class Google {
-    contentType?: string
+    params?:
+      | {
+          method: 'POST' | 'PUT';
+          data: Dict<Serializable>;
+          dataType?: 'JSON';
+        }
+      | {
+          method: 'POST';
+          data: string;
+          contentType: string;
+          dataType: 'TEXT';
+        }
+      | {
+          method: 'GET';
+          data?: UrlParams;
+        }
+      | { method: 'DELETE' },
+    progress?: ProgressCbs
   ): Promise<RT> => {
     progress = progress || {};
-    let data, url;
-    if ('upload' in progress) {
+    let url;
+    let dataPart:
+      | { method: 'POST' | 'PUT'; data: Dict<Serializable>; dataType: 'JSON' }
+      | { method: 'POST'; data: string; contentType: string; dataType: 'TEXT' }
+      | { method: 'GET'; data?: UrlParams }
+      | { method: 'DELETE' };
+    if (params?.method === 'POST' && params.dataType === 'TEXT') {
       url = `${GMAIL_GOOGLE_API_HOST}/upload/gmail/v1/users/me/${path}?uploadType=multipart`;
-      data = params;
+      dataPart = { method: 'POST', data: params.data, contentType: params.contentType, dataType: 'TEXT' };
     } else {
       url = `${GMAIL_GOOGLE_API_HOST}/gmail/v1/users/me/${path}`;
-      if (method === 'GET' || method === 'DELETE') {
-        data = params;
+      if (params?.method === 'GET') {
+        dataPart = { method: 'GET', data: params.data };
+      } else if (params?.method === 'POST' || params?.method === 'PUT') {
+        dataPart = { method: params.method, data: params.data, dataType: 'JSON' };
+      } else if (params?.method === 'DELETE') {
+        dataPart = { ...params };
       } else {
-        data = JSON.stringify(params);
+        dataPart = { method: 'GET' };
       }
     }
-    contentType = contentType || 'application/json; charset=UTF-8';
-    // eslint-disable-next-line @typescript-eslint/naming-convention
-    const headers = { Authorization: await GoogleOAuth.googleApiAuthHeader(acctEmail) };
-    const context =
-      'operationId' in progress
-        ? { operationId: progress.operationId, expectedTransferSize: progress.expectedTransferSize, frameId: progress.frameId }
-        : undefined;
-    const xhr = Api.getAjaxProgressXhrFactory('download' in progress || 'upload' in progress ? progress : {});
-    const request = { xhr, context, url, method, data, headers, crossDomain: true, contentType, async: true };
-    return (await GoogleOAuth.apiGoogleCallRetryAuthErrorOneTime(acctEmail, request)) as RT;
+    const authorization = await GoogleOAuth.googleApiAuthHeader(acctEmail);
+    // todo: const context = 'operationId' in progress ? { operationId: progress.operationId, expectedTransferSize: progress.expectedTransferSize, tabId: progress.tabId } : undefined;
+    const progressCbs = 'download' in progress || 'upload' in progress ? progress : undefined;
+    const request: Ajax = { progress: progressCbs, url, ...dataPart, headers: { authorization }, stack: Catch.stackTrace() };
```

Comment:
How about we like below?

```ts
  const baseUrl = `${GMAIL_GOOGLE_API_HOST}/gmail/v1/users/me/${path}`;
  const url = params?.method === 'POST' && 'contentType' in params
    ? `${baseUrl}?uploadType=multipart`
    : baseUrl;

  // eslint-disable-next-line @typescript-eslint/naming-convention
  const headers = { authorization: await GoogleOAuth.googleApiAuthHeader(acctEmail) };

const progressCbs = 'download' in progress || 'upload' in progress ? progress : undefined;
  // Construct request based on method and provided params
  const request: Partial<Ajax> = {
    url,
    method: params?.method ?? 'GET',
    headers,
    ...('data' in params && { data: params.data }),
    ...('contentType' in params && { contentType: params.contentType }),
    stack: Catch.stackTrace(),
    // Progress callbacks if provided
   progress: progressCbs,
  };

```

Reply:
I implemented improvements for `url` variable, but removing `dataPart` and using `data` and `contentType` directly leads to another compiler errors. Also `Partial<Ajax>` isn't compatible with `Ajax` type on `.ajax` method, will need further changes, probably can be included in https://github.com/FlowCrypt/flowcrypt-browser/pull/5372#discussion_r1385589300